src/v/io/interval_set.h

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include <absl/container/btree_set.h>
+
+namespace experimental::io {
+/**
+ * A container that contains non-empty, open intervals.
+ *
+ * Intervals that overlap with or are exactly adjacent to one another are
+ * coalesced.
+ *
+ * insert [50, 100)
+ * insert [300, 400)
+ * insert [200, 500)
+ * insert [100, 150)
+ *
+ * Result: {[50, 150), [200, 500)}
+ */
+template<std::integral T>
+class interval_set {
+    struct key {
+        // Inclusive.
+        T start;
+
+        // Exclusive.
+        T end;
+    };
+    struct compare {
+        using is_transparent = void;
+
+        bool operator()(const key& a, const key& b) const {
+            return a.start < b.start;
+        }
+
+        bool operator()(const T& a, const key& b) const { return a < b.start; }
+        bool operator()(const key& a, const T& b) const { return a.start < b; }
+        bool operator()(const T& a, const T& b) const { return a < b; }
+    };
+    using set_t = absl::btree_set<key, compare>;
+
+public:
+    using const_iterator = set_t::const_iterator;
+    using iterator = set_t::iterator;
+    struct interval {
+        T start;
+        T length;
+    };
+
+    [[nodiscard]] std::pair<const_iterator, bool> insert(interval interval);
```

Comment:
could probably use a comment about the semantics of merging

Reply:
Done